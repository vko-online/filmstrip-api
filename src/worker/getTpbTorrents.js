/* @flow */

import _ from 'lodash/fp';
import bytes from 'bytes';
import PirateBay from 'thepiratebay';

import type { Torrent } from '../types';

const bytesInGb = 1024 ** 3;
const gbToBytes = _.memoize((gb: number) => bytesInGb * gb);

const getTpbTorrents = async (
  movieTitle: string,
): Promise<Array<Torrent>> => {
  const sanitizedMovieTitle = movieTitle.replace('\'s', '');
  try {
    const results = await PirateBay.search(
      `${sanitizedMovieTitle} 1080p|720p`,
      {
        category: 207,
        orderBy: 'seeds',
        sortBy: 'desc',
      },
    );

    return _.flow(
      _.map(({ name, size, seeders, leechers, magnetLink }: {
        name: string,
        size: string,
        seeders: string,
        leechers: string,
        magnetLink: string,
      }): Torrent => ({
        source: 'The Pirate Bay',
        name,
        size: bytes(size.replace(/\s+/, '').replace('i', '')),
        seeds: parseInt(seeders, 10),
        peers: parseInt(leechers, 10),
        quality: _.cond([
          [_.includes('720p'), _.constant('720p')],
          [_.includes('1080p'), _.constant('1080p')],
          [_.stubTrue, _.constant(null)],
        ])(name),
        magnetLink,
      })),
      _.filter(_.overEvery([
        ({ name }: Torrent) => !_.includes(' HC ', name),
        _.cond([
          [
            ({ quality }: Torrent) => _.eq('720p', quality),
            ({ size }: Torrent) => size > gbToBytes(1) && size < gbToBytes(3),
          ],
          [
            ({ quality }: Torrent) => _.eq('1080p', quality),
            ({ size }: Torrent) => size > gbToBytes(2) && size < gbToBytes(5),
          ],
          [_.stubTrue, _.stubFalse],
        ]),
      ])),
      _.uniqBy('quality'),
    )(results);
  } catch (err) {
    if (_.includes('ECONNRESET', err.message)) {
      return getTpbTorrents(movieTitle);
    }

    throw err;
  }
};

export default getTpbTorrents;
