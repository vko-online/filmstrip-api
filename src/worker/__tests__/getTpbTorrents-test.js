/* @flow */

import { modelFromObject } from '../../test-utils';
import getTpbTorrents from '../getTpbTorrents';

const movieTitle = 'Pete\'s Dragon';

describe('getTpbTorrents', () => {
  it('fetches a list of torrents for a given query', async () => {
    const torrents = await getTpbTorrents(movieTitle);
    expect(modelFromObject({ torrents })).toMatchSnapshot();
  });
});
