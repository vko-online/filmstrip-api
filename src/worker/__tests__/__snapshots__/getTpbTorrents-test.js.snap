exports[`getTpbTorrents fetches a list of torrents for a given query 1`] = `
Object {
  "torrents": Array [
    Object {
      "magnetLink": "string",
      "name": "string",
      "peers": "number",
      "quality": "string",
      "seeds": "number",
      "size": "number",
      "source": "string",
    },
  ],
}
`;
